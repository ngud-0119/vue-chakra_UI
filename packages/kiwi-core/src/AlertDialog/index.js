import { createElement as h, ref } from '@vue/composition-api'
import baseProps from '../config/props'
import { forwardProps } from '../utils'
import {
  Modal,
  ModalContent,
  ModalFooter,
  ModalBody,
  ModalHeader,
  ModalOverlay,
  ModalCloseButton
} from '../Modal'

const formatIds = id => ({
  content: `alert-dialog-${id}`,
  header: `alert-dialog-${id}-label`,
  body: `alert-dialog-${id}-desc`
})

const AlertDialog = {
  name: 'AlertDialog',
  props: {
    isOpen: {
      type: Boolean,
      default: false
    },
    onClose: {
      type: Function,
      default: () => null
    },
    leastDestructiveRef: [HTMLElement, Object],
    ...baseProps
  },
  setup (props, context) {
    return () => {
      return h(Modal, {
        props: {
          isOpen: props.isOpen,
          onClose: props.onClose,
          initialFocusRef: props.leastDestructiveRef,
          formatIds,
          ...forwardProps(props)
        }
      }, context.slots.default())
    }
  }
}

const AlertDialogContent = {
  name: 'AlertDialogContent',
  props: baseProps,
  setup (props, context) {
    const innerRef = ref(null)
    return () => {
      return h(ModalContent, {
        props: {
          ...forwardProps(props)
        },
        attrs: {
          role: 'alertdialog'
        },
        ref: innerRef.value
      }, context.slots.default())
    }
  }
}

export {
  AlertDialog,
  AlertDialogContent,
  ModalOverlay as AlertDialogOverlay,
  ModalBody as AlertDialogBody,
  ModalHeader as AlertDialogHeader,
  ModalFooter as AlertDialogFooter,
  ModalCloseButton as AlertDialogCloseButton
}
