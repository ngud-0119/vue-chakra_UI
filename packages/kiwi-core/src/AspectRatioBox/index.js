import styleProps from '../config/props'
import PseudoBox from '../PseudoBox'
import { cloneVNode, forwardProps } from '../utils'

const AspectRatioBox = {
  name: 'AspectRatioBox',
  props: {
    ...styleProps,
    ratio: {
      type: Number,
      default: 4 / 3
    }
  },
  render (h) {
    const child = this.$slots.default[0]
    if (!child) return
    let vnode = cloneVNode(child, h)
    const clone = h(vnode.componentOptions.Ctor, {
      ...vnode.data,
      ...(vnode.componentOptions.listeners || {}),
      props: {
        ...(vnode.data.props || {}),
        ...vnode.componentOptions.propsData,
        position: 'absolute',
        w: 'full',
        h: 'full',
        top: 0,
        left: 0
      }
    })

    return h(PseudoBox, {
      props: {
        ...forwardProps(this.$props),
        position: 'relative',
        _before: {
          h: '0px',
          content: `""`,
          d: 'block',
          pb: `${(1 / this.ratio) * 100}%`
        }
      }
    }, [clone])
  }
}

export default AspectRatioBox
