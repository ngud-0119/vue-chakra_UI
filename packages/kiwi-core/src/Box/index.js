import { css } from 'emotion'
import __css from '@styled-system/css'
import { background, border, color, borderRadius, flexbox, grid, layout, position, shadow, space, typography, compose } from 'styled-system'
import { baseProps, propsConfig } from '../config/props'
import { forwardProps } from '../utils/'

const baseEllipsis = {
  overflow: 'hidden',
  textOverflow: 'ellipsis',
  whiteSpace: 'nowrap'
}

/**
 * @description Truncates text if `truncate` is set to true.
 * @param {Object} props Props
 */
const truncate = props => {
  if (props.truncate) {
    if (!props.lineClamp) {
      return baseEllipsis
    }
  }
}

/**
 * @description Clamps text based on number of lines.
 * @param {Object} props Props
 */
const clamp = props => {
  if (props.lineClamp) {
    return {
      ...baseEllipsis,
      '-webkit-box-orient': 'vertical',
      '-webkit-line-clamp': `${props.lineClamp}`
    }
  }
}

const decorate = props => {
  if (props.textDecoration || props.textDecor) {
    return {
      'text-decoration': `${props.textDecoration || props.textDecor}`
    }
  }
}

const system = compose(
  layout,
  color,
  space,
  background,
  border,
  borderRadius,
  grid,
  position,
  shadow,
  decorate,
  typography,
  flexbox,
  propsConfig,
  truncate,
  clamp
)

/**
 * The Box component is the base reusable component which is the building block for all other Kiwi UI components.
 * It by default renders the `<div/>`  HTMLElement.
 */
// const Box = styled('div', {
//   ...baseProps
// })`
//   ${props => {
//     const sanitizedProps = forwardProps(props)
//     return system(sanitizedProps)
//   }}
// `

const Box = {
  name: 'Box',
  inject: ['$theme'],
  props: {
    as: {
      type: String,
      default: 'div'
    },
    ...baseProps
  },
  computed: {
    theme () {
      return this.$theme()
    }
  },
  render (h) {
    const cleanedStyleProps = forwardProps(this.$props)
    const boxStylesObject = __css(system(cleanedStyleProps))(this.theme)
    const className = css(boxStylesObject)
    return h(this.as, {
      class: [className]
    }, this.$slots.default)
  }
}

export default Box
