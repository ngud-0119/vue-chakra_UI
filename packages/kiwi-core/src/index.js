import Plugin from './plugin/index.js'
export default Plugin
export { default as ThemeProvider } from './ThemeProvider'
export { default as Box } from './Box'
export { default as PseudoBox } from './PseudoBox'
export { default as Button } from './Button'
export { default as ButtonGroup } from './ButtonGroup'
export { default as IconButton } from './IconButton'
export { default as Icon } from './Icon'
export { default as Spinner } from './Spinner'
export { default as VisuallyHidden } from './VisuallyHidden'
export { default as TransitionExpand } from './TransitionExpand'
export { default as Text } from './Text'
export * from './Alert'
export * from './Avatar'
export { default as AvatarGroup } from './AvatarGroup'
export { default as Badge } from './Badge'
export { default as CloseButton } from './CloseButton'
export { default as useToast } from './Toast'
export { default as Flex } from './Flex'
export { default as Heading } from './Heading'
