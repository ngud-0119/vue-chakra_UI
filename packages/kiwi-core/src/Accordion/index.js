import { baseProps } from '../config'
import Box from '../Box'
import { forwardProps, cloneVNodes, useId } from '../utils'

const Accordion = {
  name: 'Accordion',
  props: {
    ...baseProps,
    allowMultiple: Boolean,
    allowToggle: Boolean,
    index: {
      type: Number,
      default: null
    },
    defaultIndex: {
      type: Number,
      default: 0
    }
  },
  data () {
    const initializeState = () => {
      if (this.allowMultiple) {
        return this.defaultIndex || []
      } else {
        return this.defaultIndex || 0
      }
    }
    return {
      expandedIndex: initializeState()
    }
  },
  computed: {
    isControlled () {
      return this.index != null
    },
    _index: {
      get () {
        return this.isControlled ? this.index : this.expandedIndex
      },
      set (val) {
        this.expandedIndex = val
      }
    }
  },
  methods: {
    getExpandCondition (index, itemIndex) {
      if (Array.isArray(index)) {
        return index.includes(itemIndex)
      }
      return index === itemIndex
    }
  },
  render (h) {
    const cloned = cloneVNodes(this.$slots.default, h)
    const clones = cloned.map((vnode, index) => {
      const clone = h(vnode.componentOptions.Ctor, {
        ...vnode.data,
        ...(vnode.componentOptions.listeners || {}),
        props: {
          ...(vnode.data.props || {}),
          ...vnode.componentOptions.propsData,
          isOpen: this.getExpandCondition(this._index, index)
        },
        on: {
          change: (isExpanded) => {
            if (this.allowMultiple) {
              if (isExpanded) {
                let newIndices = [...this._index, index]
                if (!this.isControlled) {
                  this.expandedIndex = newIndices
                };
                this.$emit('change', newIndices)
              } else {
                let newIndices = this._index.filter(
                  itemIndex => itemIndex !== index
                )
                if (!this.isControlled) {
                  this.expandedIndex = newIndices
                };
                this.$emit('change', newIndices)
              }
            } else {
              if (isExpanded) {
                if (!this.isControlled) {
                  this.expandedIndex = index
                };
                this.$emit('change', index)
              } else {
                if (this.allowToggle) {
                  if (!this.isControlled) {
                    this.expandedIndex = null
                  };
                  this.$emit('change', null)
                }
              }
            }
          }
        }
      }, vnode.componentOptions.children)
      return clone
    })

    return h(Box, {
      props: forwardProps(this.$props),
      attrs: {
        'data-accordion': ''
      }
    }, clones)
  }
}

const AccordionItems = {
  name: 'AccordionItems',
  props: {
    isOpen: Boolean,
    defaultIsOpen: Number,
    id: {
      type: String,
      default: useId()
    },
    isDisabled: {}
  },
  computed: {
    AccordionContext () {
      return {

      }
    }
  }
}

export {
  Accordion,
  AccordionItems
}
