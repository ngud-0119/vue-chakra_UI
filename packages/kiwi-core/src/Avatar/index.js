import { baseProps } from '../config/props'
import { forwardProps, canUseDOM } from '../utils/'
import useAvatarStyles, { avatarSizes } from './avatar.styles'
import Box from '../Box'

/**
 * @description Generate Avatar initials from name string
 * @param {String} name
 * @returns {String} Avatar Initials
 */
const getInitials = (name) => {
  let [firstName, lastName] = name.split(' ')

  if (firstName && lastName) {
    return `${firstName.charAt(0)}${lastName.charAt(0)}`
  } else {
    return firstName.charAt(0)
  }
}

/**
 * Avatar badge show's avatar status
 */
export const AvatarBadge = {
  name: 'AvatarBadge',
  inject: ['$theme', '$colorMode'],
  props: {
    size: [String, Number, Array],
    ...baseProps
  },
  computed: {
    theme () {
      return this.$theme()
    },
    colorMode () {
      return this.$colorMode()
    }
  },
  render (h) {
    const borderColorStyle = { light: 'white', dark: 'gray.800' }
    return h(Box, {
      props: {
        w: this.size,
        h: this.size,
        position: 'absolute',
        display: 'flex',
        alignItems: 'center',
        justifyContent: 'center',
        transform: 'translate(25%, 25%)',
        bottom: '0',
        right: '0',
        border: '0.2em solid',
        borderColor: borderColorStyle[this.colorMode],
        rounded: 'full',
        ...forwardProps(this.$props)
      }
    })
  }
}

/**
 * Avatar name displays an Avatar with name initials
 */
const AvatarName = {
  name: 'AvatarName',
  props: {
    name: [String, Array],
    size: [String, Array],
    ...baseProps
  },
  render (h) {
    return h(Box, {
      props: {
        w: this.size,
        h: this.size,
        textAlign: 'center',
        textTransform: 'uppercase',
        fontWeight: 'medium',
        ...forwardProps(this.$props)
      },
      attrs: {
        'aria-label': this.name
      }
    }, [this.name && getInitials(this.name)])
  }
}

/**
 * Default Avatar component shows fallback image of headshots
 */
const DefaultAvatar = {
  name: 'DefaultAvatar',
  props: {
    size: [String, Number, Array],
    ...baseProps
  },
  render (h) {
    return h(Box, {
      props: {
        h: this.size,
        w: this.size,
        lineHeight: '1rem',
        ...forwardProps(this.$props)
      },
      domProps: {
        innerHTML: `
        <svg fill="#fff" viewBox="0 0 128 128" role="img">
          <g>
            <path d="M103,102.1388 C93.094,111.92 79.3504,118 64.1638,118 C48.8056,118 34.9294,111.768 25,101.7892 L25,95.2 C25,86.8096 31.981,80 40.6,80 L87.4,80 C96.019,80 103,86.8096 103,95.2 L103,102.1388 Z" />
            <path d="M63.9961647,24 C51.2938136,24 41,34.2938136 41,46.9961647 C41,59.7061864 51.2938136,70 63.9961647,70 C76.6985159,70 87,59.7061864 87,46.9961647 C87,34.2938136 76.6985159,24 63.9961647,24" />
          </g>
        </svg>
      `
      }
    })
  }
}

/**
 * Avatar component shows images of headshots
 */
export const Avatar = {
  name: 'Avatar',
  inject: ['$theme', '$colorMode'],
  computed: {
    theme () {
      return this.$theme()
    },
    colorMode () {
      return this.$colorMode()
    }
  },
  props: {
    size: {
      type: String,
      default: 'md'
    },
    showBorder: {
      type: Boolean,
      default: true
    },
    name: [String, Array],
    src: [String, Array],
    borderColor: [String],
    ...baseProps
  },
  data () {
    return {
      image: undefined,
      hasLoaded: false
    }
  },
  created () {
    this.loadImage(this.src)
  },
  methods: {
    loadImage (src) {
      if (!canUseDOM) {
        return
      }

      const image = new window.Image()
      image.src = src

      image.onload = event => {
        this.hasLoaded = true
        this.$emit('load', event)
      }

      image.onError = event => {
        this.hasLoaded = false
        this.$emit('error', event)
      }
    }
  },
  render (h) {
    const avatarStyleProps = useAvatarStyles({
      size: this.size,
      name: this.name,
      showBorder: this.showBorder,
      borderColor: this.borderColor,
      theme: this.theme,
      colorMode: this.colorMode
    })

    const theme = this.theme
    const sizeKey = avatarSizes[this.size]
    const _size = theme.sizes[sizeKey]
    const fontSize = `calc(${_size} / 2.5)`

    /**
     * @description Render child nodes for avatar
     * @returns {Vue.VNode}
     */
    const renderChildren = () => {
      if (this.src && this.hasLoaded) {
        return h(Box, {
          props: {
            as: 'img',
            w: '100%',
            h: '100%',
            rounded: 'full',
            objectFit: 'cover',
            alt: this.name
          },
          attrs: {
            src: this.src
          }
        })
      }

      if (this.src && !this.hasLoaded) {
        if (this.name) {
          return h(AvatarName, {
            props: {
              name: this.name,
              size: _size
            }
          })
        } else {
          return h(DefaultAvatar, {
            attrs: {
              'aria-label': this.name
            }
          })
        }
      }
    }

    return h(Box, {
      props: {
        fontSize: fontSize,
        lineHeight: _size,
        ...avatarStyleProps,
        ...forwardProps(this.$props)
      }
    }, [
      renderChildren(),
      this.$slots.default
    ])
  }
}
