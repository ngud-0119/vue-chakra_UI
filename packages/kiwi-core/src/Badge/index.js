import Box from '../Box'
import { forwardProps } from '../utils'
import { baseProps } from '../config/props'
import useBadgeStyles from './badge.styles'

export default {
  name: 'Badge',
  inject: ['$theme', '$colorMode'],
  props: {
    variant: {
      type: String,
      default: 'subtle'
    },
    variantColor: {
      type: String,
      default: 'gray'
    },
    _ref: {
      type: HTMLElement
    },
    ...baseProps
  },
  render (h) {
    const badgeStyleProps = useBadgeStyles({
      theme: this.$theme(),
      colorMode: this.$colorMode,
      color: this.variantColor,
      variant: this.variant
    })

    return h(Box, {
      props: {
        d: 'inline-block',
        textTransform: 'uppercase',
        fontSize: 'xs',
        fontFamily: 'body',
        px: 1,
        rounded: 'sm',
        fontWeight: 'bold',
        whiteSpace: 'nowrap',
        verticalAlign: 'middle',
        ...badgeStyleProps,
        ...forwardProps(this.$props)
      }
    }, this.$slots.default)
  }
}
