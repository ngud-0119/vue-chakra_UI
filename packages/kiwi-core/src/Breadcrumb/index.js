import { ref, createElement as h } from '@vue/composition-api'
import { baseProps } from '../config/props'
import Box from '../Box'
import Link from '../Link'
import { forwardProps, cloneVNodes } from '../utils'

const BreadcrumbSeparator = {
  name: 'BreadcrumbSeparator',
  props: {
    spacing: [String, Number, Array],
    separator: [String, Object],
    ...baseProps
  },
  setup (props) {
    const innerRef = ref(null)

    return () => {
      return h(Box, {
        props: {
          as: 'span',
          mx: props.spacing,
          ...forwardProps(props)
        },
        attrs: {
          role: 'presentation'
        },
        ref: innerRef.value
      }, [props.separator])
    }
  }
}

const Span = {
  name: 'Span',
  props: {
    ...baseProps
  },
  setup (props, context) {
    const innerRef = ref(null)

    return () => {
      return h(Box, {
        props: {
          as: 'span',
          ...forwardProps(props)
        },
        ref: innerRef.value
      }, context.slots.default())
    }
  }
}

const BreadcrumbLink = {
  name: 'BreadcrumbLink',
  props: {
    isCurrentPage: Boolean,
    ...baseProps
  },
  setup (props, context) {
    const innerRef = ref(null)
    const Comp = props.isCurrentPage ? Span : Link

    return () => {
      return h(Comp, {
        props: {
          ...forwardProps(props)
        },
        attrs: {
          'aria-current': props.isCurrentPage ? 'page' : null
        },
        ref: innerRef.value
      }, context.slots.default())
    }
  }
}

const BreadcrumbItem = {
  name: 'BreadcrumbItem',
  props: {
    isCurrentPage: Boolean,
    isLastChild: Boolean,
    separator: [Object, String],
    addSeparator: Boolean,
    spacing: [String, Number, Array],
    ...baseProps
  },
  setup (props, context) {
    return () => {
      const children = context.slots.default().filter(e => e.tag)
      const clones = cloneVNodes(children, h).map((clone) => {
        if (clone.componentOptions.tag === BreadcrumbLink.name) {
          const { propsData } = clone.componentOptions
          propsData['isCurrentPage'] = props.isCurrentPage
          clone.componentOptions.propsData = propsData
          return clone
        }
        if (clone.componentOptions.tag === BreadcrumbSeparator.name) {
          const { propsData } = clone.componentOptions
          propsData['spacing'] = props.spacing
          propsData['separator'] = props.separator
          clone.componentOptions.children = clone.componentOptions.children || props.separator
          clone.componentOptions.propsData = propsData
          return clone
        }
      })

      return h(Box, {
        props: {
          display: 'inline-flex',
          alignItems: 'center',
          as: 'li'
        }
      }, [
        ...clones,
        !props.isLastChild && props.addSeparator && h(BreadcrumbSeparator, {
          props: {
            spacing: props.spacing,
            separator: props.separator
          }
        })
      ])
    }
  }
}

const Breadcrumb = {
  name: 'Breadcrumb',
  props: {
    spacing: {
      type: [String, Number, Array],
      default: 2
    },
    addSeparator: {
      type: Boolean,
      default: true
    },
    separator: {
      type: [String, Object],
      default: '/'
    },
    ...baseProps
  },
  setup (props, context) {
    return () => {
      const children = context.slots.default().filter(e => e.tag)
      const clones = cloneVNodes(children, h)
        .map((node, index, array) => {
          const { propsData } = node.componentOptions
          propsData['addSeparator'] = props.addSeparator
          propsData['separator'] = props.separator
          propsData['spacing'] = props.spacing
          propsData['isLastChild'] = array.length === index + 1
          node.componentOptions.propsData = propsData
          return node
        })

      return h(Box, {
        props: {
          as: 'nav',
          ...forwardProps(props)
        },
        attrs: {
          'aria-label': 'breadcrumb'
        }
      }, [h(Box, {
        props: {
          as: 'ol'
        }
      }, clones)])
    }
  }
}

export {
  BreadcrumbSeparator,
  BreadcrumbLink,
  BreadcrumbItem,
  Breadcrumb
}
