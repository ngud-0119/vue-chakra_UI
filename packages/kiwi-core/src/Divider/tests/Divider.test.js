import Divider from '../'
import { render } from '@/packages/kiwi-core/src/test-utils'

const renderComponent = (props) => {
  const inlineAttrs = (props && props.inlineAttrs) || ''
  const base = {
    components: { Divider },
    template: `<Divider ${inlineAttrs} />`,
    ...props
  }
  return render(base)
}

it('should render correctly', () => {
  const { asFragment } = renderComponent()

  expect(asFragment()).toMatchSnapshot()
})

it('should change orientation', () => {
  const inlineAttrs = 'orientation="vertical"'
  const { asFragment } = renderComponent({ inlineAttrs })

  expect(asFragment()).toMatchSnapshot()
})
