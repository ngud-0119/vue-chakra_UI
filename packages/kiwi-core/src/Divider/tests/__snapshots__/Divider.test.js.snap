// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should change orientation 1`] = `
<DocumentFragment>
  <hr
    aria-orientation="vertical"
    class="css-1n8n8cf"
  />
</DocumentFragment>
`;

exports[`should render correctly 1`] = `
<DocumentFragment>
  <hr
    aria-orientation="horizontal"
    class="css-boncrs"
  />
</DocumentFragment>
`;
